collect_app/src/test/java/org/odk/collect/android/http/mock/MockHttpClientConnection.java

```diff
@@ -0,0 +1,68 @@
+package org.odk.collect.android.http.mock;
+
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import org.odk.collect.android.http.HttpCredentialsInterface;
+import org.odk.collect.android.http.HttpGetResult;
+import org.odk.collect.android.http.HttpHeadResult;
+import org.odk.collect.android.http.OpenRosaHttpInterface;
+import org.odk.collect.android.utilities.ResponseMessageParser;
+
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.HttpURLConnection;
+import java.net.URI;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class MockHttpClientConnection implements OpenRosaHttpInterface {
+
+    private boolean getHttpShouldReturnNull;
+
+    @NonNull
+    @Override
+    public HttpGetResult get(@NonNull URI uri, @Nullable String contentType, @Nullable HttpCredentialsInterface credentials) throws Exception {
+        if (getHttpShouldReturnNull) {
+            return null;
+        }
+
+        String xml =
+        "<forms>" +
+        "<form url=\"https://opendatakit.appspot.com/formXml?formId=CascadingSelect\">Cascading Select Form</form>" +
+        "<form url=\"https://opendatakit.appspot.com/formXml?formId=widgets\">Widgets</form>" +
+        "<form url=\"https://opendatakit.appspot.com/formXml?formId=NewWidgets\">New Widgets</form>" +
+        "<form url=\"https://opendatakit.appspot.com/formXml?formId=sample\">sample</form>" +
+        "</forms>";
+
+        InputStream is = new ByteArrayInputStream(xml.getBytes());
+
+        Map<String, String> headers = new HashMap<>();
+        headers.put("X-OpenRosa-Version", "1.0");
+        headers.put("Content-Type", "text/xml;charset=utf-8");
+
+        return new HttpGetResult(is, headers, "test-hash", HttpURLConnection.HTTP_OK);
+    }
+
+    @NonNull
+    @Override
+    public HttpHeadResult head(@NonNull URI uri, @Nullable HttpCredentialsInterface credentials) throws Exception {
+        return new HttpHeadResult(0, new HashMap<String, String>());
+    }
+
+    @NonNull
+    @Override
+    public ResponseMessageParser uploadSubmissionFile(@NonNull List<File> fileList, @NonNull File submissionFile, @NonNull URI uri, @Nullable HttpCredentialsInterface credentials) throws IOException {
+        return null;
```

Comment:
How difficult is it for us to mock this behaviour and test it? When we do the switch to `okHttp` are there tests related to upload functionality that verify it's behaviour?

Reply:
We could mock this method and get the pretty complex logic in InstanceServerUploader.uploadSubmissionFile() tested. However, this class is mocking the OpenRosaHttpInterface so it won't necessarily help test that okhttp is working correctly.

MockedServerTest could possibly help with writing an androidTest that would test okhttp upload.

Reply:
Okay! No problem!